compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -2266,110 +2300,152 @@ static void setLoweringConfigForComputeOps(func::FuncOp entryPointFn,
-          // consistent vector tile sizes.
-          for (auto i : llvm::seq<int64_t>(
-                   0, std::min(tileAndFuseSizes.size(), vecTileSizes.size()))) {
-            if (tileAndFuseSizes[i])
-              vecTileSizes[i] = 0;
-          }
+  LLVM_DEBUG(KD_DBGS() << "Parallel vector tile sizes: " << parallelVecTileSizes
+                       << "\n");
+
+  // Split parallel vector tile sizes into common parts and op-specific parts.
+  SmallVector<int64_t> commonVecTileSizes = parallelVecTileSizes;
+  SmallVector<int64_t> innerVecTileSizes(maxLoopNums, 0);
+  for (auto op : computeOps) {
+    auto iterTypes = cast<TilingInterface>(op).getLoopIteratorTypes();
+    for (auto [idx, iterType] : llvm::enumerate(iterTypes)) {
+      if (iterType == utils::IteratorType::reduction) {
+        innerVecTileSizes[idx] = parallelVecTileSizes[idx];
+        commonVecTileSizes[idx] = 0;
+      }
+    }
+  }
 
-          SmallVector<int64_t> reductionTiles;
-          splitParallelAndReductionTiles(genericOp, vecTileSizes,
-                                         reductionTiles);
-          setVectorSizesForDynamicShapes(genericOp, vecPreProcStrategy,
-                                         vecTileSizes, reductionTiles);
+  // Set the lowering configs with new tile sizes.
+  for (auto op : computeOps) {
+    int numLoops = cast<TilingInterface>(op).getLoopIteratorTypes().size();
+    TileSizesListType tileSizesList;
+    ScalableTileFlagsListType scalableTileFlagsList;
+
+    // For root op, we patch the adjusted tile sizes on its original tiling
+    // config.
+    if (op == rootOperation) {
+      tileSizesList = rootLoweringConfig.getTileSizeVals();
+      scalableTileFlagsList = rootLoweringConfig.getScalableTileFlagVals();
+      if (tilingConfig.getNumTilingLevels() > 0) {
+        tileSizesList[tilingConfig.getDistributionLevel()] = distTileSizes;
+      }
+      if (tilingConfig.getNumTilingLevels() > 1) {
+        tileSizesList[tilingConfig.getVectorCommonParallelLevel()] =
+            commonVecTileSizes;
+      }
+    } else {
+      // Build 4-level lowering configs for other ops.
+      tileSizesList = {distTileSizes, commonVecTileSizes};
+      SmallVector<int64_t> zeros(numLoops, 0);
+      TypeSwitch<Operation *>(op)
+          .Case<tensor::PackOp>([&](auto packOp) {
+            tileSizesList.push_back(zeros);
+            tileSizesList.push_back(innerVecTileSizes);
+            // Scale and permutate the outer dim tiles for pack op.
+            ArrayRef<int64_t> innerTiles = packOp.getStaticInnerTiles();
+            ArrayRef<int64_t> dimPos = packOp.getInnerDimsPos();
+            auto outerDimsPerm = packOp.getOuterDimsPerm();
+            for (auto &tileSizes : tileSizesList) {
+              for (auto [pos, size] : llvm::zip_equal(dimPos, innerTiles)) {
+                if (tileSizes[pos] == 0 || ShapedType::isDynamic(size))
+                  continue;
+                tileSizes[pos] = tileSizes[pos] / size;
```

Comment:
[note] I think it's not guaranteed that lowering configs match tiled ops in this case. See the below scaling (which is from adjustTileSieForPackOp). Though that's probably fine because pack op eventually get decomposed to pad+reshape+transpose ops.

```
  // Scale to actual tile sizes with the pack op's inner tile sizes.
  for (auto [pos, size] : llvm::zip_equal(innerDimsPos, innerTiles)) {
    vecTileSizes[pos] *= size;
  }
  for (auto [pos, size] : llvm::enumerate(vecTileSizes)) {
    ...
    // If other ops already set a smaller tile size, don't override it to avoid
    // too large tile size on them.
    parallelVecTileSizes[pos] = std::min(parallelVecTileSizes[pos], size);
```

I had a concern about turning generic vectorizer to look into lowering_config first. Then I understood that it's for scalable vectors. I'm not pushing back the approach. Mostly wanna flag that something is off in pack op codegen using scalable vectors.

Reply:
I don't think this is only for scalable vectors. The use of the `ValueBounds` analysis to get the vector sizes was a gas stopper for when `lowering_config` couldn't encode the information that we needed for vectorization. This analysis itself has its limitations and shouldn't override `lowering_config`, which is where we purposely set the tile size information. Now that we propagate `lowering_config` to more ops within the dispatch, we have to make sure that the tile size information carried on it is correct because passes are free to look at `lowering_config` and use that information. IOW, if we propagate the `lowering_config` from one op to another one we have to make sure that the information we are propagating is correct for the recipient op. Otherwise, we shouldn't be propagating it.